src/bls12_381/ec.rs

```diff
@@ -130,18 +138,61 @@ macro_rules! curve_impl {
                     let mut y2 = self.y;
                     y2.square();
 
-                    let mut x3b = self.x;
-                    x3b.square();
-                    x3b.mul_assign(&self.x);
-                    x3b.add_assign(&Self::get_coeff_b());
-
-                    y2 == x3b
+                    y2 == $affine::y2_from_x(self.x)
                 }
             }
 
             fn is_in_correct_subgroup_assuming_on_curve(&self) -> bool {
                 self.mul($scalarfield::char()).is_zero()
             }
+
+            fn sw_encode(mut t: $basefield) -> $projective {
+                use ::LegendreSymbol::*;
+
+                // handle the case t == 0
+                if t.is_zero() { t = Self::get_swenc_const0() };
+
+                // w = (t^2 + b + 1)^(-1) * sqrt(-3) * t
+                let mut w = t;
+                w.square();
+                w.add_assign(&Self::get_coeff_b());
+                w.add_assign(&$basefield::one());
+                // handle the case t^2 + b + 1 == 0
+                if w.is_zero() { return $projective::zero() };
+                w = w.inverse().unwrap();
+                w.mul_assign(&Self::get_swenc_const0());
+                w.mul_assign(&t);
+
+                // x1 = - wt  + (sqrt(-3) - 1) / 2
+                let mut x1 = w;
+                x1.mul_assign(&t);
+                x1.negate();
+                x1.add_assign(&Self::get_swenc_const1());
+                let alpha = Self::y2_from_x(x1).legendre();
+
+                // x2 = -1 -x1
+                let mut x2 = x1;
+                x2.negate();
+                x2.sub_assign(&$basefield::one());
+                let beta = Self::y2_from_x(x2).legendre();
+
+                // x3 = 1/w^2 + 1
+                let mut x3 = w;
+                x3.square();
+                x3 = x3.inverse().unwrap();
+                x3.add_assign(&$basefield::one());
+
+                let x = match (alpha, beta) {
+                    (QuadraticResidue, _) => x1,
+                    (_, QuadraticResidue) => x2,
+                    _ => x3
+                };
+
+                let mut y = Self::y2_from_x(x).sqrt().unwrap();
+                if t.legendre() == QuadraticNonResidue { y.negate() }
+                let p = Self {x: x, y: y, infinity: false};
+                p.scale_by_cofactor()
```

Comment:
Since the results of two different points `a,b` being sw_encoded are being summed together, this is currently computing `cofactor * sw_encode(a) + cofactor * sw_encode(b)`. A cofactor multiplication could be saved by instead doing `cofactor * (sw_encode((a) + sw_encode((b))`.

Reply:
oh that's a good point - I'm not sure how the library will behave when adding points not in the subgroup but that's definitely something that should be done.

thanks!